javalib/src/main/scala/java/util/concurrent/ConcurrentSkipListSet.scala

```diff
@@ -0,0 +1,404 @@
+package java.util.concurrent
+
+import java.util._
+import java.lang.Comparable
+import scala.collection.mutable
+import scala.math.Ordering
+
+import scala.collection.JavaConversions._
+
+class ConcurrentSkipListSet[E >: Null] protected (ordering: Ordering[E], provided: Boolean) 
+    extends AbstractSet[E]
+    with NavigableSet[E]
+    with Cloneable
+    with Serializable { self =>
+
+  def this() =
+    this(ConcurrentSkipListSet.defaultOrdering[E], false)
+
+  def this(comparator: Comparator[E])  = 
+    this(Ordering.comparatorToOrdering(comparator), true)
+
+  def this(sortedSet: SortedSet[E]) = {
+  	this(Ordering.comparatorToOrdering(sortedSet.comparator()), true)
+    addAll(sortedSet)
+  }
+
+  def this(collection: java.util.Collection[E]) = {
+  	this(ConcurrentSkipListSet.defaultOrdering[E], false)
+    addAll(collection)
+  }
+
+  def this(comparator: Comparator[E], collection: Collection[E]) = {
+    this(Ordering.comparatorToOrdering(comparator), true)
+    addAll(collection)
+  }
+
+  implicit object BoxOrdering extends Ordering[Box[E]] {
+    def compare(a: Box[E], b:Box[E]): Int = ordering.compare(a.inner, b.inner)
+  }
+
+  private val inner: mutable.SortedSet[Box[E]] = new mutable.TreeSet[Box[E]]()
+
+  override def isEmpty(): Boolean =   
+    inner.isEmpty
+
+  def iterator(): Iterator[E] = {
+    new Iterator[E] {
+      private val iter = inner.iterator
+
+      var actual: Option[E] = None
+
+      def hasNext(): Boolean = iter.hasNext
+
+      def next(): E = {
+        actual = Some(iter.next().inner)
+        actual.get
+      }
+
+      def remove(): Unit = actual.map(self.remove(_))
+    }
+  }
+
+  override def add(e: E): Boolean =
+    if (e === null) throw new NullPointerException()
+    else inner.add(Box(e))
+
+  override def addAll(c: Collection[_ <: E]): Boolean = {
+    val iter = c.iterator()
+    var changed = false
+    while (iter.hasNext)
+      changed = add(iter.next()) || changed
+    changed
+  }
+
+  override def clear(): Unit =
+    inner.clear() 
+
+  override def clone(): ConcurrentSkipListSet[E] =
+    new ConcurrentSkipListSet(this)
+
+  def comparator(): Comparator[E] = {
+    if (!provided) null
+    else  {
+      new Comparator[E] {
+        def compare(a: E, b: E) = ordering.compare(a, b)
+      }
+    }
```

Comment:
Oh you're absolutely right to return `null`. I was talking about the non-null case. If the user *does* provide a comparator in the constructor, this method will return a *different* comparator. Its `compare` method will do the same thing as the original one, but otherwise it won't be *the same*. That's what I'm concerned about.